File: src/scripts/check_release_includes_discovery.py
#!/usr/bin/env python3
"""Assert dist/discovery/discovery-manifest.json ships with the package.

Phase 5.3 of R3 — wired as a `prepublishOnly` hook in package.json and
re-used by the publish workflow. Fails loudly when:
  - dist/discovery/discovery-manifest.json is missing
  - the file is empty or not valid JSON
  - the artefacts array is empty
  - the summary sibling is missing (release tarball ships both)

The intent is that `npm publish` (and `npm pack`) refuse to produce a
silently-broken artifact where the discovery contract surface
(ADR-013) is absent from the consumer install path.
"""
from __future__ import annotations

import json
import sys
from pathlib import Path

ROOT = Path(__file__).resolve().parents[2]
MANIFEST = ROOT / "dist" / "discovery" / "discovery-manifest.json"
SUMMARY = ROOT / "dist" / "discovery" / "discovery-manifest.summary.md"


def _die(msg: str) -> int:
    sys.stderr.write(f"check-release-discovery: {msg}\n")
    sys.stderr.write(
        "  hint: run `python3 scripts/build_discovery_manifest.py --write --strict`"
        " before `npm pack` / `npm publish`.\n"
    )
    return 1


def main() -> int:
    if not MANIFEST.is_file():
        return _die(f"{MANIFEST.relative_to(ROOT)} is missing.")
    raw = MANIFEST.read_text(encoding="utf-8").strip()
    if not raw:
        return _die(f"{MANIFEST.relative_to(ROOT)} is empty.")
    try:
        data = json.loads(raw)
    except json.JSONDecodeError as exc:
        return _die(f"{MANIFEST.relative_to(ROOT)} is not valid JSON: {exc}.")
    artefacts = data.get("artefacts")
    if not isinstance(artefacts, list) or not artefacts:
        return _die(
            f"{MANIFEST.relative_to(ROOT)} carries no artefacts — discovery"
            " scanner produced an empty manifest."
        )
    if not SUMMARY.is_file():
        return _die(f"{SUMMARY.relative_to(ROOT)} is missing.")
    sys.stdout.write(
        f"check-release-discovery: OK ({len(artefacts)} artefacts in"
        f" {MANIFEST.relative_to(ROOT)})\n"
    )
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: src/scripts/check_release_includes_discovery.ts
#!/usr/bin/env tsx
/**
 * Assert dist/discovery/discovery-manifest.json ships with the package.
 *
 * TypeScript twin of `src/scripts/check_release_includes_discovery.py`
 * (ADR-200, Phase 4 / Wave 4c). Mirrors the Python CLI contract EXACTLY —
 * no flags, exit codes (0 OK, 1 die), stdout/stderr split, byte-identical
 * finding messages (paths relative to ROOT), same checks in the same order.
 * No behaviour changes.
 *
 * Phase 5.3 of R3 — wired as a `prepublishOnly` hook in package.json and
 * re-used by the publish workflow. Fails loudly when:
 *   - dist/discovery/discovery-manifest.json is missing
 *   - the file is empty or not valid JSON
 *   - the artefacts array is empty
 *   - the summary sibling is missing (release tarball ships both)
 */

import * as fs from 'node:fs';
import * as path from 'node:path';
import { fileURLToPath, pathToFileURL } from 'node:url';

const _HERE = fileURLToPath(import.meta.url);
// src/scripts/check_release_includes_discovery.ts → two dirs up is repo root.
// Mirrors Python `Path(__file__).resolve().parents[2]`.
const ROOT = path.resolve(path.dirname(_HERE), '..', '..');
const MANIFEST = path.join(ROOT, 'dist', 'discovery', 'discovery-manifest.json');
const SUMMARY = path.join(ROOT, 'dist', 'discovery', 'discovery-manifest.summary.md');

/** POSIX relative path under ROOT (mirrors `Path.relative_to(ROOT)`). */
function _rel(target: string): string {
    return path.relative(ROOT, target).split(path.sep).join('/');
}

function _isFile(p: string): boolean {
    try {
        return fs.statSync(p).isFile();
    } catch {
        return false;
    }
}

function _die(msg: string): number {
    process.stderr.write(`check-release-discovery: ${msg}\n`);
    process.stderr.write(
        '  hint: run `./scripts-run src/scripts/build_discovery_manifest --write --strict`' +
            ' before `npm pack` / `npm publish`.\n',
    );
    return 1;
}

function main(): number {
    if (!_isFile(MANIFEST)) {
        return _die(`${_rel(MANIFEST)} is missing.`);
    }
    const raw = fs.readFileSync(MANIFEST, 'utf-8').trim();
    if (raw === '') {
        return _die(`${_rel(MANIFEST)} is empty.`);
    }
    let data: unknown;
    try {
        data = JSON.parse(raw);
    } catch (exc) {
        const msg = exc instanceof Error ? exc.message : String(exc);
        return _die(`${_rel(MANIFEST)} is not valid JSON: ${msg}.`);
    }
    const artefacts =
        data !== null && typeof data === 'object'
            ? (data as Record<string, unknown>)['artefacts']
            : undefined;
    if (!Array.isArray(artefacts) || artefacts.length === 0) {
        return _die(
            `${_rel(MANIFEST)} carries no artefacts — discovery` +
                ' scanner produced an empty manifest.',
        );
    }
    if (!_isFile(SUMMARY)) {
        return _die(`${_rel(SUMMARY)} is missing.`);
    }
    process.stdout.write(
        `check-release-discovery: OK (${artefacts.length} artefacts in` +
            ` ${_rel(MANIFEST)})\n`,
    );
    return 0;
}

const _isCliEntry =
    process.argv[1] !== undefined &&
    import.meta.url === pathToFileURL(path.resolve(process.argv[1])).href;
if (_isCliEntry || process.argv[1] === _HERE) {
    process.exit(main());
}

export { ROOT, MANIFEST, SUMMARY, main };
